### PR TITLE
Fix S3904 FP: Exclude Razor View assemblies

### LINKED ISSUE #2921: S3904 false positive with razor and shared assemblyInfo file
### Description

When having a razor project with a AssemblyInfo file we get 

warning S3904: Provide an 'AssemblyVersion' attribute for this assembly.

We started seeing this at about the same time from FxCop and it since has been patched.  May be a similar issue.

https://github.com/dotnet/roslyn-analyzers/issues/2143

We want to used a shared AssemblyInfo file among our projects. This is why we have it configured this way.

### Repro steps

1. Create a project for razor

dotnet new razor -n RazorSonarQube  

2. Edit the csproj
<GenerateAssemblyInfo>false</GenerateAssemblyInfo>

3. Add a AssemblyInfo.cs file with the following contexts

[assembly: AssemblyCompany("CompanyName")]
[assembly: AssemblyProduct("CompanyName")]
[assembly: AssemblyCopyright("Copyright (C) CompanyName")]
[assembly: AssemblyVersion("1.0.0.0")]

4. Add SonarAnalyzer.CSharp nuget package

5. dotnet clean

6. dotnet build

7.  We get the warning.

### Related information

* SonarC# Version: 8.1.0.13383
* Visual Studio Version: 16.4.2
* Dot Net: 3.1

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #2921

Changes needed in .NET ITs:
* fixed target files to get issues for Razor generated assemblies
* configure `NetCore31` to analyze generated code, added Razor Views
* fixed temporary path handling to deterministic version

### CODE REVIEW

File: analyzers/its/SonarAnalyzer.Testing.ImportBefore.targets [hunk 1]
```diff
@@ -43,6 +43,7 @@
     <PropertyGroup Condition="$(SonarAnalyzerRegressionTestRunning)">
       <!-- Set the path to the SARIF report -->
       <ErrorLog>$(MSBuildStartupDirectory)\output\$(PROJECT)\$(AssemblyName)-$(ProjectGuid).json</ErrorLog>
+      <RazorCompilationErrorLog>$(MSBuildStartupDirectory)\output\$(PROJECT)\$(AssemblyName)-$(ProjectGuid).Views.json</RazorCompilationErrorLog>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/its/SonarAnalyzer.Testing.ImportBefore.targets:46]: Reason for this is explained here https://github.com/SonarSource/sonar-scanner-msbuild/issues/949

File: analyzers/its/create-issue-reports.ps1 [hunk 2]
```diff
@@ -108,10 +108,12 @@ function GetActualIssues([string]$sarifReportPath){
         return
     }
 
-    # Change spaces to %20
+    # Change spaces to %20 and replace temporary paths
+    $tempFileNameRegex = '^.*\.(NETFramework|NETCoreApp),Version='
     $allIssues.location | Foreach-Object {
         if ($_.uri) {
             $_.uri = $_.uri.replace(' ', '%20')
+            $_.uri = [System.Text.RegularExpressions.Regex]::Replace($_.uri, $tempFileNameRegex, 'Replaced-Temporary-Path\.$1,Version=')
```
Comment (pavel-mikula-sonarsource) [inline analyzers/its/create-issue-reports.ps1:116]: I had to remove this hack from `regression-test.ps1`, because it first sorted the issues by URI (so the temp path was sometimes before normal path and sometimes it wasn't) and then did the replace. Actual temp filename information was lost on the way.

New version does the replacement before sorting to make it deterministic. And it also preserves the actual temp file name (replaces only the temp directory) so we can see what file is it from (like `Replaced-Temporary-Path\.NETCoreApp,Version=v3.1.AssemblyAttributes.cs`).

File: analyzers/its/expected/NetCore31/NetCore31--S1128.json [hunk 3]
```diff
@@ -0,0 +1,56 @@
+{
```
Comment (costin-zaharia-sonarsource) [inline analyzers/its/expected/NetCore31/NetCore31--S1128.json:1]: The file name `NetCore31--S1128.json` seems to be missing the project GUID. Does the script works well? I remember that the scanner complains if the project does not have a GUID.

File: analyzers/its/expected/NetCore31/NetCore31--S1858.json [hunk 4]
```diff
@@ -1,17 +0,0 @@
-{
-"issues":  [
-{
-"id":  "S1858",
-"message":  "There's no need to call 'ToString()' on a string.",
```
Comment (costin-zaharia-sonarsource) [inline analyzers/its/expected/NetCore31/NetCore31--S1858.json:5]: Why was this removed?

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!
